### ISSUE TITLE
Restoring without local database takes ages

### ISSUE THREAD

**Issue by jarmo** [duplicati/duplicati#1391]
I tried to restore a single ~500kB file from a ~500GB backup without a local database. It takes hours and hours and i can see a lot of `Downloading file (27,22 KB) ...` messages. As soon as i specify local database with `--dbpath` then the restore is done in minutes.

It does take hours even if the backup destination is local disk.

There seems to be something fundamentally wrong when restoring a single file with exact path from local destination takes more than few minutes.

Why does it take that long?

Using latest duplicati 2.x version.
##

**Comment by jarmo** [comment]
It finally finished. Pay attention to the duration.

```
Checking remote backup ...
  Listing remote folder ...
Searching backup 0 (25.05.2015 1:44:00) ...
Checking existing target files ...
  1 files need to be restored (563,18 KB)
Scanning local files for needed data ...
1 remote files are required to restore
  Downloading file (49,95 MB) ...
  0 files need to be restored (0 bytes)
Verifying restored files ...
Restored 1 (563,18 KB) files to c:\Users\Jarmo\Desktop\restore
Duration of restore: 15:24:03
```

**Comment by jarmo** [comment]
Restoring now all files from a 70 and 700 GB backup without local database files. Wish me luck, because this is a real situation this time and not just a test. Currently seeing "Downloading file (47.78 MB) ..." appearing REALLY slowly although the backup location is a local disk.

**Comment by kenkendk** [comment]
If you do not have a local database, Duplicati will build one before continuing.
There is some logic in the latest (2.0.0.90) version that checks if it can avoid to restore the entire database, which should speed things up. If something is not matching correctly, Duplicati will start looking for the needed information in potentially all files.

If you need to perform the restore, I recommend that you run the "repair" command first to generate the local database. This will make it possible to stop and resume the restore process without needing to generate the database again.

The 2.0.0.90 version also features a recovery tool, which works with a simple text file instead of a database.

**Comment by jarmo** [comment]
@kenkendk thank You for your reply.

I can now tell more about the whole situation, described in the order of steps done:

1) HDD failed on 15th Oct;
2) Made clean install of Windows 10 at Oct 16th;
3) Started two simultaneous restore processes at Oct 16th of ~70GB and ~700GB backups - both without having a local database present;
4) At the evening of Oct 18th I saw that restore process for ~70GB was successfully completed, but there were no files restored from my ~700GB backup due to error similar to #1143 - not a single file - same problem can be reproduced on a Windows and Linux environment;
5) Fortunately I had a local databases backupped in that ~70GB backup for both of the backups so I started restore process again of ~700GB backup, this time with the option `--dbpath` option pointed to the recently restored sqlite database - this happened on the same evening of Oct 18th;
6) Today is Oct 28th and restore process has still not (!!!) finished - there's still about ~140GB to restore according to duplicati;

Good thing is that I've managed to get everything back from the first backup and it seems that restore process for the larger backup has been slowly progressing. Can't still say for sure if I've managed to get everything back yet or not since I'm not touching that PC just in case.

There's some things I've noticed:
* `--no-local-db=true` does not seem to make any difference when trying to restore without a database, but reading from the `advanced` help section it does seem that rebuilding of the local database is skipped;
* I have quite a good performance desktop PC with 8 CPU cores, but duplicati restore process only utilizes about ~12% of that CPU constantly;
* There's almost no IO according to Performance Monitor or even HDD led's - yes there are occasional spikes, but most of the time nothing happens;
* Restoring from the GUI (e.g. web-page) does not work (at least, didn't restore anything within a 48h from that ~70GB backup) - it seems that backups as big are just not suitable for that GUI solution;

In conclusion:
* In some cases, Duplicati won't even restore a single file due to checksum mismatch, when trying to restore without a database - however, when restoring with a database, checksum does match - weird;
* Duplicati does not use CPU nor IO power fully to perform restore operation, thus making it really painfully slow;

Some background about the contents of my backups as well:
* ~70GB backup is from a system disk containing mostly applications configuration, documents, etc - e.g. many small files;
* ~700GB backup consists mostly of RAW images, psd and jpeg images - more bigger files;

In short - restore process is still in progress (after 10+ days!), but at least there's hope for getting stuff back. However, I have to admit that I'm quite disappointed already in Duplicati and looking for (even paid) alternatives.

Hopefully, these issues will be resolved in a future version of duplicati and no-one else have to go through similar suffering.

**Comment by jarmo** [comment]
Update: restore process finally finished.

Here's what it outputted:
```
Restored 50641 (766,85 GB) files to h:/restore
Duration of restore: 04:39:52
```

I'm not sure what it means by 4 hours, because it actually happened from 18th Oct to 1st Nov.

**Comment by kenkendk** [comment]
The duration text displays hh:mm:ss, so the additional days are not shown.

Happy to hear that it worked after all that time.

I am aware of performance issues, but if you see no disk IO, I would guess that the majority of the time is spent on network traffic (either to a local share, or to the remote end).

**Comment by jarmo** [comment]
@kenkendk, that isn't the case - there was no network IO present as well and there would not have been any need (unless everything is sent to NSA :P) because backup was on a different (from restore disk) local physical disk. In other words, there was no need to retrieve a single byte over a network.

**Comment by kenkendk** [comment]
I assume H:\ was a network mount, but if it is a physically attached disk then there is no reason there should be network IO.

I would guess that time is spent creating transaction snapshots for the SQLite database, but that should show up as either CPU usage or disk usage.

Do you have any guess as to where the time goes?

**Comment by jarmo** [comment]
Yeah, it is a physical disk, not a network mount. I don't have any guesses, it seemed as if duplicati didn't know that I have more than one core on my CPU or the `--thread-priority` is always very low.

I myself used `--thread-priority=high` as a setting, but it didn't seem to make any difference without specifying that option at all.

I also used `--debug-output=true --log-file="h:/duplicati-restore.log" --log-level="Information"` as options, but I guess `Information` level is too high. I could try to restore again with lower log level.

The version I'm used for restoring was `2.0.0.87_preview_2015-07-03`, maybe newer version is better?

Can you give me any command line options which might help you more?

**Comment by jarmo** [comment]
I've cloned this repository and ran `Duplicati.Commandline` with debugger using VS Express.

Of course I didn't wait until restore process finished (it would have been taken more than 14 days with debugger, I suppose :P), but here's some findings.

* It seems that most of the time is spent in SQlite databases;
* Initial loading process before actual restoring takes place, takes quite much time. Here's few examples:

```
- GetFilesAndSourceBlocksFast ~ 275s
- // Fill BLOCKS with remote sources
  var volumes = database.GetMissingVolumes().ToList();
  if (volumes.Count > 0) ~469s, volumes.Count=15 000
```

But since this happens only (at least that's how it seemed to me), maybe this is not that a big of a problem.

It seems that a lot of time is spent within restore loop:

```
- foreach(var restorelist in volumekeeper.FilesWithMissingBlocks) ~ 35s
- foreach(var restoremetadata in volumekeeper.MetadataWithMissingBlocks) ~ 26s
- blockmarker.UpdateProcessed(result.OperationProgressUpdater) ~10s
- blockmarker.Commit(result) ~7s
```

In other words, spending time in SQlite for every block (50MB?) takes at least an extra minute and since I have 15 000 files (volumes.Count) to process then it takes roughly ~10 days.

Looking at the code, it seems that many complex select statements are performed with multiple sub-queries etc. and none of the (at least) temporary tables have any indexes.

Maybe the problem isn't just the lack of indexes, but there's also some SQlite locks involved?

Anyway, I bet that all this problem exists only for large backups and doesn't happen in usual development environment.

I can't investigate this problem any further right now and I don't feel myself quite home with VS, but I hope that some of this information helps.

**Comment by jarmo** [comment]
Maybe here's something useful for optimizing queries in SQlite https://www.sqlite.org/optoverview.html

**Comment by JF28** [comment]
Hi,

I encounter exactly the same problem : restoration takes ages without the database.

I configured a backup with Duplicati (v2.0.0.87) between two servers. Duplicati sends its zipped files from the server A to server B via SFTP. No problem for the backup itself, it works smoothly daily (size of data to back up = 160GB divided in 30000 files).

Problem arises when I want to restore files on server B (therefore without the database, which is located on server A). It seems that Duplicati literally wastes lots of time just to rebuild the database. In my last attempt, I left Duplicati run for two days on server B. After two days, the database was still not completely rebuilt, and therefore absolutely no file could be restored in 48 hours. Crazy !

Now, if I manually sends the database from server A to server B via SFTP, and rerun Duplicati on server B with the proper argument --dbpath, files restoration immediately starts at an acceptable speed.

So, I have two questions for Kenneth.

1) Why does it take so long to rebuild the database on server B ? dblocks files are available on local high-end hard disks (6Gb/s SAS), so there is nothing to download on network. CPU (Xeon E5-2440v2, 8 cores) is not a concern either, nor RAM (32GB). Duplicati has plenty of resources available to rebuild the database, but uses very little of it, and it seems to get bogged down in endless SQL queries (profiling log just shows countless SELECT queries, with few INSERT statements).

2) As a workaround, is it not conceivable that Duplicati saves his SQLite database on the backend at the end of each backup, so that the database is immediately available if a restoration must be performed on a computer other than the one that made the backup ?

Thank you !

**Comment by albertgouws** [comment]
The way I'm doing my backup is to create my entire backup on the server
being backed up, including the local DB in the same folder.

Once the backup is completed, I have a robocopy script run which takes all
that and copies it to my backup location (including the local DB).

When I restore, I specify the local DB which resides on the backup location
and everything works reasonably quickly....

Perhaps something to look at doing, always copy across the local DB to the
backup location and then when restoring:


   1. Check local location for Local DB
   2. Check backup location for Local DB
   3. Rebuild Local DB



On Mon, Nov 9, 2015 at 9:07 AM JF28 <notifications@github.com> wrote:

> Hi,
>
> I encounter exactly the same problem : restoration takes ages without the
> database.
>
> I configured a backup with Duplicati (v2.0.0.87) between two servers.
> Duplicati sends its zipped files from the server A to server B via SFTP. No
> problem for the backup itself, it works smoothly daily (size of data to
> back up = 160GB divided in 30000 files).
>
> Problem arises when I want to restore files on server B (therefore without
> the database, which is located on server A). It seems that Duplicati
> literally wastes lots of time just to rebuild the database. In my last
> attempt, I left Duplicati run for two days on server B. After two days, the
> database was still not completely rebuilt, and therefore absolutely no file
> could be restored in 48 hours. Crazy !
>
> Now, if I manually sends the database from server A to server B via SFTP,
> and rerun Duplicati on server B with the proper argument --dbpath, files
> restoration immediately starts at an acceptable speed.
>
> So, I have two questions for Kenneth.
>
> 1) Why does it take so long to rebuild the database on server B ? dblocks
> files are available on local high-end hard disks (6Gb/s SAS), so there is
> nothing to download on network. CPU (Xeon E5-2440v2, 8 cores) is not a
> concern either, nor RAM (32GB). Duplicati has plenty of resources available
> to rebuild the database, but uses very little of it, and it seems to get
> bogged down in endless SQL queries (profiling log just shows countless
> SELECT queries, with few INSERT statements).
>
> 2) As a workaround, is it not conceivable that Duplicati saves his SQLite
> database on the backend at the end of each backup, so that the database is
> immediately available if a restoration must be performed on a computer
> other than the one that made the backup ?
>
> Thank you !
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/duplicati/duplicati/issues/1391#issuecomment-154861948>
> .
>
-- 

Regards,
Albert

www.gouws.org | blog.gouws.org | www.flickr.com/albert_gouws

**Comment by jarmo** [comment]
This is not exactly the same - mine took ages even with the database. As I already wrote, there seem to be problems with (lack of) indexes in that database.

Also, exposing database to the 3rd location includes slight privacy risk - all paths with filenames are in the database. It's not a problem when not using encryption and privacy does not matter that much.

I have plans to look more into that problem, but don't know when I get time to do that yet.

**Comment by kenkendk** [comment]
@jarmo Thanks for the measurements. I have not optimized the queries for the restore process, but I will take a look at the problematic ones that you discovered.

The `FilesWithMissingBlocks` and `MetadataWithMissingBlocks` queries are almost identical, so they can be optimized in one go. The `UpdateProcessed` query is a count, which I think thank be extracted from the actual update queries so that time goes away completely.

The `Commit` call is likely not possible to address, as that is what writes the database to disk.

**Comment by kenkendk** [comment]
I made a new build with some added indexes: http://updates.duplicati.com/preview/duplicati-2.0.0.91_preview_2015-11-18.zip

Could you try it out, preferably with the debugger attached so we can see if the indexes helped?

**Comment by verybadsoldier** [comment]
I am also experiencing very very slow restore performance here. I am using Windows 8.1 x64 machine to restore my 2 GB backup from Google Drive. After 1 hour it had restored about 84 MB only. So at that pace it would take about 24 hours to restore the whole thing.
Also repairing the database did not make a difference in performance. Any help for me?

It says "You are currently running Duplicati 2.0.0.91- 2.0.0.91_preview_2015-11-18

**Comment by verybadsoldier** [comment]
It is also slow when restoring from a local storage backup instead of Google Drive. So it really seems to have nothing to do with the internet connection but solely with CPU usage for some reason.

**Comment by mach-o** [comment]
Same issue here.

I backed up 58GB from a Ubuntu laptop to a remote Raspberry Pi device, and then tried to locally restore the backup, as a trial run for the scenario where my laptop suffers some unfortunate mishap.

The "No local database, building a temporary database" phase took around 22 hours, then "Searching backup 0" took around 5 hours and (rightly) reported about ½ million files to restore. The restore has now been stuck on "Scanning local files for needed data..." for about two days. Although the directory structure has been restored, no files have yet been.

This is using the git master version from about five days ago, which reports itself as "Version: 2.0.0.7 - DEBUG". I admit to being largely ignorant of Duplicati's internal workings, and the Pi is certainly not a  processing powerhouse, but it's not clear to me why unzipping a copy of the backup would need to take this long. I backed up with --no-encryption, which I would assume should also speed things up.

**Comment by JF28** [comment]
Just for information, my own feedback relative to my own situation (see my previous message). Even with the latest version of Duplicati (git-master), restoration from a purely *local* backend (again, see my previous message) takes a *huge* time. Even worse : I said previously that avoiding the complex rebuild process of the database by backing up the sqlite file on the backend was a good workaround to speed the restoration up tremendously. In fact, I was compeltely wrong. It worked just once. All restore tests I did subsequently have been failures, with or without database rebuilding : each restoration test seems just to sink in the mud (I have every time abandoned after several days of waiting, with virtually no file restored). I'm in the same situation as jarno : restoration takes ages, even with the database.

Sadly, it seems that Duplicati can not handle "large" backup (consisting of tens of gigabytes of data).

**Comment by mach-o** [comment]
Thanks for that information, @JF28. I have the luxury of being able to let my restore attempt run indefinitely, so I will try to write back here if it ever finishes (or progresses).

**Comment by verybadsoldier** [comment]
@JF28 
For me it does not even seem to be related to huge backups: As said above, restoring just 2GB of data takes about 24 hours for me (local db). Does anyone know maybe an alternative to Duplicati? This restore problem is quite a disadvantage for me.

**Comment by kenkendk** [comment]
The "Scanning local files ..." is checking to see if any of the files on your disk has the blocks needed, so it can avoid downloading remote data. You can disable this with `--no-local-blocks`.

The restore process can be optimized in many places, but I have not gotten around to looking at it just yet.

As for alternatives, I think this is as close to a definitive list as possible (although the Duplicati entry is not updated):
https://en.wikipedia.org/wiki/List_of_backup_software

**Comment by mach-o** [comment]
@verybadsoldier : The best alternative will depend on your particular needs and platform(s). I personally want free software that's straightforward to setup and that makes encrypted, scheduled, differential backups on Linux, so helping to improve Duplicati to get it there seems my best option. I've disqualified all the other tools that I've seriously considered, namely because:
- [rsync](https://en.wikipedia.org/wiki/Rsync), though otherwise attractive and elegant, isn't designed for differential backups;
- [duplicity](https://help.ubuntu.com/community/DuplicityBackupHowto) was crashy even at its initial steps when I tried to use it;
- [backuppc](http://backuppc.sourceforge.net/) isn't designed for encrypted backups, and seems aimed more at larger network administration than personal use; and
- [bacula](http://www.bacula.org/) has a monumental learning curve that I am very happy not to have to subject myself to.

**Comment by mach-o** [comment]
After running the restore I mentioned above for about four days, extrapolating the full restore time based on the progress indicator suggests that it will take about four years to complete. Oof! Looks like it's back to rsync for me for the moment.

**Comment by kenkendk** [comment]
Ouch!

It will be a while before I can look at the restore speed.

Do you have any idea where the time is used? How many files? How big is the backup? What kind of disk and CPU are you using?

**Comment by mach-o** [comment]
58GB in ½ a million files, using an ARM11 processor (Raspberry Pi) on an external USB drive (Western Digital). Not sure where the time is used.

Incidentally, I should correct my backup software comparison above, by noting that rsync is, in fact, designed to do differential backups, using its --link-dest option.

**Comment by FootStark** [comment]
Hi @kenkendk ,

restore process is still awfully slow.
I found out that the update on progress takes several seconds.

When using an index with (FileId, Index) on the update table, it works fine. It sped up my restore from 2 hours to 3 minutes (250 MB).

I'm new to GitHub, so I post the change here. Hope you can incorperate it soon.

LocalRestoreDatabase.cs:Line_769
`m_updateTable = "UpdatedBlocks-" + Library.Utility.Utility.ByteArrayAsHexString(Guid.NewGuid().ToByteArray());`
`m_insertblockCommand.ExecuteNonQuery(string.Format(@"CREATE TEMPORARY TABLE ""{0}"" (""FileID"" INTEGER NOT NULL, ""Index"" INTEGER NOT NULL, ""Hash"" TEXT NOT NULL, ""Size"" INTEGER NOT NULL, ""Metadata"" BOOLEAN NOT NULL)", m_updateTable));`
`m_insertblockCommand.ExecuteNonQuery(string.Format(@"CREATE INDEX ""{0}_FileIdIndexIndex"" ON ""{0}"" (""FileId"", ""Index"")", m_updateTable));`
`// not necessary: m_insertblockCommand.ExecuteNonQuery(string.Format(@"CREATE INDEX ""{0}_HashSizeIndex"" ON ""{0}"" (""Hash"", ""Size"")", m_updateTable));`
`// not necessary: m_insertblockCommand.ExecuteNonQuery(string.Format(@"CREATE INDEX ""{0}_IndexIndex"" ON ""{0}"" (""Index"")", m_updateTable));`

**Comment by kenkendk** [comment]
Great find!!

I will make sure to merge it in. Is the Hash-Size index not being used?

**Comment by kenkendk** [comment]
It looks like it is used in line 842, but perhaps it has no real impact on the overall time.

**Comment by kenkendk** [comment]
Also, there is another performance fix in issue #1569 which reduces the time to rebuild the local database.

**Comment by JF28** [comment]
Can't wait the next preview release ^____________________________^

**Comment by FootStark** [comment]
Index Hash-Size might or might not be used, but i guess it is not. The Index on Index is toxic, because it is most likely no index at all. If I got it right, Index denominates the number of a block in a file. If that is true, it is almost always single digit (FileSizes < 1M). Thus, there is no distinctiveness and any optimization attempt on this results in a forced nested loop.
Depending on how sophisticated the stats and execution plan optimization of SQLite is (awareness of index discintiveness?), I guess it tries to optimize Joins using Index-Index, because it is smaller (single field, Int) and thus appears better suited than Hash-Size. I can't prove that, as I don't know how to retrieve actual execution plans from SQLite (I work with MsSql most of my time).
Anyway, Index on FileId and Index is much better suited, as it is small (2 Ints) and thus fast, very disctincitve (unique?), represents the order the rows are added (table is actually clustered), and logically represents the primary join condition to the blocks table (as used in UpdateProcessed and the Update on Commit). This is why this index is sufficient and I removed the other 2.

**Comment by kenkendk** [comment]
Great, thanks for clarifying!
I just wanted to make sure there was a reason, and what you say makes perfect sense.

In case you are interested in the SQLite execution plan, you can get it with `EXPLAIN`:
https://www.sqlite.org/lang_explain.html

Now that you mention MsSQL, do you see a potential benefit from being able to connect to MsSQL? 
(Not suggesting removing SQLite, but rather to allow alternate RDMS systems).

**Comment by kenkendk** [comment]
@JF28: Here you go: http://updates.duplicati.com/preview/duplicati-2.0.0.99_preview_2016-02-15.zip

**Comment by kenkendk** [comment]
@jarmo, @verybadsoldier, @mach-o: can you retry your setups, and see if we are within a tolerable timeframes now?

**Comment by FootStark** [comment]
Hmm, from the top of my head, I think SQLite is a solid choice for this application. There are not many benefts to leverage from other DBMS. Some pros and cons:

Pro others:
- Better multithreading Support
- Better concurrency engines (Client-Server)
- More sophisticated optimizations
- Better development environments, Easy (external) Performance monitoring

Con others:
- Additional installations necessary
- More Overhead and complexity in Code
- Additional sources for errors

I do not see, where the Pros do actually apply here (e.g. concurrent db Access). From my perspective, because of the rather small datasets of Duplicati (in DBMS terms), the cons outweigh the Advantages by far. It would be better to fully leverage the possibilities with SQLite (e.g. allow to backup database file after main backup is done, solves any rebuild issues). Small Performance issues like above can be tracked and removed as necessary.

tl:dr; Stick with SQLite solely.

**Comment by kenkendk** [comment]
@FootStark: Thanks!

**Comment by FootStark** [comment]
@kenkendk  Thank you for pointing me to `EXPLAIN`, it's very useful!
Out of curiousity i put it to a test on the problem at hand, and it actually proves that IndexIndex was used prior to the fix (line 4). That means future results should be consistently faster.

```
0	0	0	SEARCH TABLE Blocks-04DFA3B70EAB5247B6FC3946AE24CF59 USING INDEX Blocks-04DFA3B70EAB5247B6FC3946AE24CF59_RestoredMetadataIndex (Restored=?)
0	0	0	EXECUTE LIST SUBQUERY 1
1	0	0	SCAN TABLE Blocks-04DFA3B70EAB5247B6FC3946AE24CF59
1	1	1	SEARCH TABLE UpdatedBlocks-5F63D042E260E249B6AF4221AA192477 USING INDEX UpdatedBlocks-5F63D042E260E249B6AF4221AA192477_IndexIndex (Index=?)
0	0	0	SEARCH TABLE Blocks-04DFA3B70EAB5247B6FC3946AE24CF59 USING INTEGER PRIMARY KEY (rowid=?)
```

**Comment by verybadsoldier** [comment]
Great stuff guys! Just restored 3.4 GB from Google Drive within ~4 minutes! Big thanks!

**Comment by mach-o** [comment]
Great news, I'll try it when I have a minute!

**Comment by jarmo** [comment]
I've been trying to restore with existing local database using version from `master` and it has restored `~5.5GB out of 760GB` in a hour. Calculating roughly it will take about `10 days` to restore it all. In other words, it seems to be slightly faster (I'm also running it in Visual Studio Express using debug mode with all breakpoints disabled, but it might still affect performance).

I just got an idea, which might (or not) improve performance speed even more - what if we delete the "used" rows from temporary tables so that the tables will get smaller after each file restored? Would that idea even work in general?

Although, I'm not too familiar with inner workings of Duplicati, and I have only around `50000` files to be restored (which is not that big of an amount really) so maybe there are still better places to optimize.

When I have more time, I might look into it again in more detail, but can't see it happening in the near future.

I will leave the restore process running for about 24h to see results after that.

PS! I'm using exactly the same backup with exactly the same local database as the real problem happened - I left these untouched just in case I need to do something like this I'm doing right now.

**Comment by jarmo** [comment]
Ok, it has been restoring about 21 hours now and it's fully restored ~200GB. Calculating from that, I can get a rough estimate of about 4 days or slightly less.

I'd say this is much better for such a big backup :)

Although, there's still much room for improvement (e.g. if you compare it to just with copying that amount of data).

Still, big thank you to everyone involved :)

**Comment by FootStark** [comment]
@jarmo Thank you for sharing your test results. Can you ellaborate a little on the terms of your test? To be precise: What storage was involved for Source, Backup and Target (local disk, USB, Online, ...)? Was there source data still available (duplicati uses the original source files if they are accessible), already target data present, or was everything downloaded (restored from backup files)?
There should be a (rather large) log file `Duplicati.debug.log` in your bin folder. Can you sanatize (remove filenames) and share it to reproduce where the time was used?
Also, you may want to consider running another test with the release build preview version, that would deliver a better measure for achievable speed.

**Comment by jarmo** [comment]
@FootStark I have already written about in great detail starting from first comment in this thread or even more important comment here https://github.com/duplicati/duplicati/issues/1391#issuecomment-148520207.

TLDR; I lost all data and had to perform full restore from local physical (SATA 2) hard drive to another (SATA 2) hard drives. In other words, no network was involved.

At first restore process was impossible, because rebuilding local database failed due to some checksum errors, but later I managed to get a local database backup so I could use `--dbpath` option, which resulted about 14 days of restoration process. I did not use PC at the time and it did not use a lot of resources according to resource monitor (only about 12% of total CPU, and no single thread/core was maxed out).

What do you mean that using release build preview version would deliver a better speed than using version from `master`? Do you mean that debugger will slow it down so much?

**Comment by FootStark** [comment]
I did read it, but I was not sure what your current setup is. Now that the data is back I did not assume you deleted it all again.
So the following could apply: Assume Backup-Location is E:\BCK, Original data location was D:\Data.
Last time (real desaster) you restored all your data to the original folders (D:\Data) where they were backupped from earlier. This time you tried to restore in an alternate folder (e.g. D:\Test), as the original folder already exists. In this case Duplicati would copy data between the original and restored files (D:\Data --> D:\Test), and not restore it from the backup in E:\BCK (except modified data). And that's a whole different area of code.
Regarding the debugger: It slows down execution only moderately as long as no conditional breakpoints are used. But in debug mode and without optimization it could still be somewhere at only half speed to release. For a real world measurement on processing speed, it should not be used. The goal is to see what is really achievable right now. And there something like 10 - 20 MB/sec would be good (at least for how I would want to use it).

**Comment by jarmo** [comment]
Okay, to clarify - I know that Duplicati tries to use existing blocks already so I moved everything away from these folders so Duplicati would think that it needs to restore everything. That's how I got similar setup as losing really everything as happened before. I will give it a go with precompiled binary at some point too to see if there's much of a difference.

**Comment by kenkendk** [comment]
I doubt the debugging makes much difference. It has very low impact in my measurements. The SQLite version can have a huge impact.

The `--no-local-blocks` option disables reuse of local data, so combined with `--no-local-db` it should be the same as restore-from-scratch.

I know there is at least one potential optimization in the restore: restore blocks in zip order instead of db order. This will cause read-ahead and other things to work much better. I am sure there are other database optimizations as well.

**Comment by FootStark** [comment]
OK, I might take another look at the restore process on the weekend. Gotta build a suitable large backup for testing first. My current one is too small, and it contains a big chunk of executable data, which produces a heavy workload for malware protection. Which leads to another question: maybe restore files with a work suffix (e.g. .restoring) and rename all in the end?

**Comment by kenkendk** [comment]
@FootStark Would you restore to a temp folder and then move in, or just the individual files?

There are pros and cons to using a temp folder.
Pros:
 - No interference to programs using the files
 - No chance for files being accessed in partial state
 - No programs will update the files during the restore

Cons:
 - Extra time spent for moving
 - Extra space used to store files (if originals exist)
 - Needs to copy data from existing files (if they exist)
 - Needs to re-apply metadata after moving (to preserve timestamps)

But if it solves in issue for you, I would be happy to include it.

**Comment by FootStark** [comment]
I think the best solution would be to rename the files within their folders and do the restore in-place. For small backups or single file restores Temp-Folder Solutions are ok. For larger ones like whole drives rather not. 

Process could be: 
1. Find all files that will have to be touched
2. Rename the files if already existing (e.g. *.restoring, maybe just use hashcode as Name)
3. Restore the files as usually in place
4. As soon as a file is done, verify and rename to original.

I am not sure if metadata has to be restored twice then because of rename, but maybe it can be read before the final renaming and be restored immediatly afterwards.

Pros of this solution would be:
- No extra space needed
- (hopefully) only minimal changes to current restore process
- Still prevents others from touching files still in restoration
- Files completely done can be easily identified and already be used before full completion
- Executables would (hopefully) be prevented from mutlible inspections by Virus/malware scanner during restore.

No real cons there. Only ones I could think of:
- Maybe special treatment of long filepathes necessary (for very heavily nested file structures, just appending an .restoring might result in too Long pathes) Yet unlikely.


What do you thinik? Maybe we should move to another issue for discussion ?

BTW: I found another optimization in restore, speeding up very large restores like jarmo's. I will try to check in to github tonight.

**Comment by JF28** [comment]
Hi all,

Here is my feedback with the last preview version (2016-02-15). I restarted from scratch: I deleted all my old backups and previous (unfinished) restore tests.

So, after making a fresh complete backup (~173GB,  ~34000 files), I tried a full restore, as previously, directly from a local backend (hosted on high-quality SAS hard drives => no need to download anything from network). Below, an excerpt from the debug log generated during restoration :

```
2016-02-17 11:52:13Z - Information: The operation Restore has started
2016-02-17 11:52:13Z - Information: No local database, building a temporary database
2016-02-17 12:16:10Z - Profiling: Recreate temporary database for restore took 00:23:54.992
2016-02-17 12:16:14Z - Information: Needs to restore 33966 files (172,35 GB)
2016-02-17 12:22:05Z - Information: 2683 remote files are required to restore

(following lines just to show you the download speed from my sas drives :))
2016-02-17 12:22:06Z - Profiling: Downloaded 49,97 MB in 00:00:00.7968959, 62,70 MB/s
2016-02-17 12:22:13Z - Profiling: Downloaded 49,96 MB in 00:00:00.1562593, 319,72 MB/s
2016-02-17 12:22:31Z - Profiling: Downloaded 49,97 MB in 00:00:00.3437227, 145,39 MB/s
2016-02-17 12:22:45Z - Profiling: Downloaded 49,94 MB in 00:00:00.1562491, 319,59 MB/s
2016-02-17 12:22:54Z - Profiling: Downloaded 49,95 MB in 00:00:00.1406018, 355,25 MB/s

2016-02-17 20:30:48Z - Profiling: Starting - RestoreVerification
2016-02-17 21:47:35Z - Profiling: RestoreVerification took 01:16:46.143
```

So :

- database rebuild took ~23 minutes (still a little bit slow, but it's a huge improvement !)
- files restoration took ~8 hours (same remark)
- restore verification took ~77 minutes (too long imho)

All in all, Duplicati is now much better. Kudos to Kenneth & FootStark :)

(And I completely agree with FootStary on temporary filenames during restoration)

**Comment by FootStark** [comment]
I found a way to speed up file-restoration to pretty much full disk IO (well, single thread decryption is the limiting factor now). @kenkendk: Before I check in, is there a reason I cant see why BlockMarker writes to a temp table before commiting to Blocks (Setting Restored to 1)? If not, I would commit directly. This has no impact on the speed, but allows for accurate statistics during restore.

**Comment by kenkendk** [comment]
@FootStark: I've added issue #1586 to discuss the temp-file scheme.

I will see if I can figure out the reason for the "restored" flag.

**Comment by kenkendk** [comment]
The "UpdatedBlocks" (temp) table keeps track of what file chuncks are written. The `restored=1` is set once all copies of that block has been written to all files.

I cannot see why this should be impossible to do without writing this information, but instead just update the block table. It may have to do with choosing the order, such that you get all copies of a specific block sequentially, instead of ordering by filename.

It looks like it should be possible to set `restored=1` directly after each write of a block without problems, which would also bypass the problem.

And full disk-speed sounds extremely nice!

**Comment by JF28** [comment]
I tested a new restore scenario. This time it was an incremental restore : I still had full data previously restored on 2016-02-17, and I expected that Duplicati use these data to restore only files that have changed since 2016-02-17, in much less time than my first restore test (see my previous post). I knew that few files had changed, since the second incremental backup sent only 800MB of zipped files on the backend (first full backup sent 130GB of zipped files). Excerpt from the debug log :

```
2016-02-19 14:11:39Z - Information: The operation Restore has started
2016-02-19 14:11:39Z - Information: No local database, building a temporary database
2016-02-19 14:40:03Z - Profiling: Recreate temporary database for restore took 00:28:21.828
2016-02-19 14:40:08Z - Information: Needs to restore 34072 files (172,59 GB)
2016-02-19 14:41:52Z - Profiling: Starting - ScanForexistingTargetBlocks
2016-02-19 21:42:18Z - Profiling: ScanForexistingTargetBlocks took 07:00:26.368
2016-02-19 21:42:18Z - Profiling: Starting - ScanForExistingSourceBlocksFast
2016-02-19 21:44:38Z - Profiling: ScanForExistingSourceBlocksFast took 00:02:19.953
2016-02-19 21:45:34Z - Information: 69 remote files are required to restore
2016-02-19 22:00:02Z - Profiling: Starting - RestoreVerification
2016-02-19 23:07:47Z - Profiling: RestoreVerification took 01:07:45.025
```

In summary :

- database rebuild took ~29 minutes
- scan for existing data on target folder took ~**7 hours** (awfully long !)
- files restoration took ~15 minutes (as expected, very few data have changed)
- restore verification took ~78 minutes (why is it taking so long ? What does Duplicati exactly at this level ?)
- total restoration time : ~9 hours (relatively poor performance given the real amount of data to be restored)

**Comment by FootStark** [comment]
@JF28 It's same database issue in the restore process causing this. I already have a fix, I will streamline it and check in soon. Just wait a couple of days for a new version. 
For verification: Duplicati checks and hashes ALL files from the backup. That 78 mins are your CPU-limit on reading an hashing files, as it is single threaded. This can only be improved by parallelization.

**Comment by kenkendk** [comment]
@JF28 in addtion to FootStark's comments, you can use the `--skip-restore-verification` option to skip the  restore verification part if you prefer. The scanning for local blocks can be skipped with `--no-local-blocks`, but it should not take that long. I think the patch from FootStark will address this.

I am working on a concurrent design that will make it simple to exploit multithreaded hashing. It could start verification in parallel with the restore itself.

**Comment by JF28** [comment]
Thanks for your explanations & work, guys :)

Kenneth : if I skip the local blocks scanning, Duplicati will restore every bit of the backup contents, even if 95% of the data are already restored and up-to-date in the target folder, right ? The aim of my second restore test was precisely to speed up the restore process by using existing data from a previous (successful) restore, and only overwriting/altering the few files that have changed since this previous restore (let's call this an "incremental restore"). Correct me if I'm wrong, but it just can not be done with `--no-local-blocks` enabled ?

**Comment by kenkendk** [comment]
@JF28 I misread the log.

`ScanForexistingTargetBlocks` checks if the data is already present in the destination files, and cannot be disabled.
`ScanForExistingSourceBlocksFast` checks for required blocks in *other* files, and can be disabled with `--no-local-blocks`.
I can see that the time spent in your setup is actually the check to see if data is already present in the files, so there is no quick-fix to improve this.
Lets see what happens with @FootStark's fix... I have high hopes!

**Comment by FootStark** [comment]
I am pretty sure it will. It targets two operations being very slow - thats finding missing blocks and the progress statistics.
I am currently having trouble to track all my changes, I am new to github and have another primary source control solution. I should be able to clean my code and commit on the weekend.

**Comment by mach-o** [comment]
I finally got around to another benchmark here, and things are similarly much improved. This might be the wrong thread in which to raise this issue, but I did notice a striking anomaly related to download speed.

I backed up ~30MB in 423 files to Amazon S3 with encryption and compression from my laptop, which generated a 13½MB backup in about 2½ minutes.

Restoring to my laptop with --no-local-blocks --no-local-db to a new --restore-path took 40 seconds. Does that represent a rebuild from scratch? Seems hard to believe it could be that fast.

A fully db-less restore to the external USB drive attached to my  raspberry pi device took 12 minutes, 51 seconds: a huge improvement over the earlier results I was getting.

This is also where the anomaly comes in. though. It was hard to tell exactly, but my sense was that more than half of the restore time (well over 6 minutes) was taken up by the download phase ("Downloading file (13.50 MB) ..."). When I `wget` the same 13.50MB file directly from Amazon S3 to the raspberry pi, it takes only 25 seconds. Is it possible that the S3 download is missing a significant potential optimization?

**Comment by FootStark** [comment]
@mach-o Were you running duplicati on the pi? This might have to do something with temp-files, which are likely to be stored on the flash drive. Maybe a new `temp-in-memory` option should be considered, which should work on the pi too, if blockfiles are not too large (maybe use 20MB instead of 50MB default).

**Comment by mach-o** [comment]
@FootStark yes; to the pi, with an externally powered USB hard drive.

**Comment by FootStark** [comment]
@JF28 , on your incremental restore scenario (or the prior backup), did you use `--skip-metadata` option? I'm experiencing almost all files being downloaded, even when nothing has to be restored. This seems to be the metadata handling, and metadata seems to be scattered all over the block files...?

**Comment by kenkendk** [comment]
Metadata should be placed in the same dblock file as the content. But... if the metadata is shared by multiple files, then it might be somewhere else.

One consideration could be to hold back small files (and metadata) and attempt to group them into a single volume. This would most likely make it faster to restore a set of small files, but could mean that you need to download two volumes if you restore a file larger than the threshold for a "small" file.

@mach-o you can use `--tempdir` to set the folder where the downloads are placed (and other stuff is stored temporarily). If you set it to point at the USB disk, you should see improved speed if the flash drive is the problem. Also, the download process is also decrypting the files, which might take some time too.

**Comment by JF28** [comment]
No, I didn't use `--skip-metadata` in any scenario

**Comment by FootStark** [comment]
There is a new [experimental build out now](https://github.com/duplicati/duplicati/releases/tag/v2.0.1.7-2.0.1.7_experimental_2016-03-19) that incorporates many changes that should - amongst other great features - improve speed for large restores drastically (details [here](http://www.duplicati.com/news/majorupdatecoming)).
Besides internal query optimizations now cpu-intensive cryptographic operations can be speed up by multithreading using options `enable-piped-streaming` and `aes-set-threadlevel=1..4`.
File verification was also improved by not re-verifying files in the end that were verified on the initial file scan in the beginning.
Feedback on the performance in actual user installations (@jarmo, @JF28) would be appreciated to find out if speed is sufficient now.

**Comment by JF28** [comment]
Hi FootStark,

I downloaded version 2.0.1.7 (meanwhile, I see that Kenneth released a newer version 2.0.1.8, too bad) and launched a new full restore test on an empty target folder, without the database, and with the new options `--enable-piped-streaming` (I expect that this option changes nothing in my environment, as my backend is purely local, and the bandwidth of my hard drives is clearly not an issue) and `--aes-set-threadlevel=4`. Preliminary results are not good : I launched Duplicati 4 hours ago, and so far, not a single byte of data has yet been restored. From what I see in the huge profiling log (1.5GB), Duplicati is still in the process of rebuilding the database... Back in the nightmare :(

I leave Duplicati running, I'll keep you informed of the progress.

PS : I made a whole bunch of backups since [February 18](https://github.com/duplicati/duplicati/issues/1391#issuecomment-185879872) : my backend now contains 23 backup sets, the Duplicati database is significantly larger (~ 1.3GB) than when I made my first restore attempt on February 18 (size of the database, with a single backup set, was then ~350MB, if I recall correctly), so I fully expected that the rebuilding of the database will take more time... But 4 hours of waiting to have nothing at all, it's a lot of wasted time :(

**Comment by FootStark** [comment]
Hi JF28, first of all thank you for testing.
Despite this issue's title I actually did not look into the recreation process, but solely into the core of file restoration. Right now it would be great to find out if that is sufficiently fast, I will then look if db-recreation can be speed up, too.
I hope you can let Duplicati finish building the database so there will be a benchmark for later tests. If it's not done in some hours, you might want to try a restore with local db for now. But please keep the log-file for further analysis, if the few GB's do not hurt you to much... A small duration summary from the log would be great when done as you posted above for your last tests.
Furthermore to track recreation, please take a look at recurring entries in the log that comprise a cycle in the recreation process. If you could provide such a block, it would help to see where the times are lost...

**Comment by JF28** [comment]
Glad that my very tiny contribution can help you :) I'll keep the log of course, I can send it to you in a 7-zip archive later if you want. You can certainly learn more from it than me :)

Anyway, Duplicati is still in the middle of... something. Not a single file yet recovered. But it's very, very, very strange : Duplicati is currently downloading a bunch of dblock files. I can't remember that Duplicati needed dblock files for the rebuilding of the database, peviously. So, I made a comparison between the log of today and the log of my last successful restore, two weeks ago. I think I found something interesting. Here is an excerpt of the log dated March 6 :

```
2016-03-06 19:37:37Z - Information: The operation Restore has started
2016-03-06 19:37:37Z - Information: No local database, building a temporary database
2016-03-06 19:37:39Z - Profiling: Starting - Recreate temporary database for restore
2016-03-06 20:04:26Z - Profiling: Starting - CommitRecreatedDb
2016-03-06 20:05:10Z - Profiling: CommitRecreatedDb took 00:00:44.610
2016-03-06 20:58:46Z - Profiling: Recreate temporary database for restore took 01:21:06.337
```

As you can see, during the build of temporary database, there is an operation "CommitRecreatedDb", then other relatively long operations (I don't know what exactly, but the log contains only a few entries between 20:05 and 20:58 : just 2 INSERT statements and 6 SELECT queries), and voila, the database is ready (which is clearly indicated in the log by the sentence _Recreate temporary database for restore took XXX_).

Now, excerpt from today :

```
2016-03-20 16:02:29Z - Information: The operation Restore has started
2016-03-20 16:02:29Z - Information: No local database, building a temporary database
2016-03-20 16:02:31Z - Profiling: Starting - Recreate temporary database for restore
2016-03-20 16:39:11Z - Profiling: Starting - CommitRecreatedDb
2016-03-20 16:40:31Z - Profiling: CommitRecreatedDb took 00:01:19.172
2016-03-20 16:40:53Z - Information: Processing required 660 blocklist volumes
2016-03-20 20:49:23Z - Information: Processing all of the 2158 volumes for blocklists
```

The last two entries are completely new to me, and I can't found any equivalent entries in previous restore logs that I kept. Strange. Do you understand what Duplicati is doing ? I guess in any case that the database is still not ready, because I still don't see "Recreate temporary database took ..." in the log, and no files are being restored currently.

**Comment by FootStark** [comment]
Hmm, AFAIK Duplicati downloads the blockfiles if it is not happy with the information it could gather from the list files. There've been some things fixed recently, and there is an addtional fix in 2.0.1.8, so maybe it's already repaired.
And yes, until it says "Recreate temporary database took xxx", it is not done with this step.
Regarding your 23 filesets: Are these just 23 incremental backups of the same folder, or did you perform additional backups (other source folders) into the same backup target folder?
Yet this still takes too long. so the log would be great. You can send it [here](mailto:use.0ab1.once@lueck-senses.de) (or a link to it if it is too large when zipped).

**Comment by JF28** [comment]
It's 23 incremental backups (selection unchanged between each backup).

I sent you the log, good luck :)

**Comment by FootStark** [comment]
Thanks, I will look into it. First impression: db restore will run another 5-6 hours maybe. So you might want to cancel and give it a try with the local db.

**Comment by kenkendk** [comment]
The two new log entries happen because Duplicati cannot find the required information in the index files.
It then looks into the dblock files where they should be found (the count 600), and then there appears to be more missing, so it starts to grab the rest of the dblock files.

A new feature is that after the repair completes, any new information will be re-uploaded such that the next repair will not need to do this.

I did rewrite some of the queries related to recreating the database, such that they are more resilient against wrongly stored data. I do not know the performance impacts of this rewrite on your set, but I did not see any significant changes in my test runs.

There should not be any changes that causes it to need more data than previous versions, so that could indicate a new problem.

The 2.0.1.8 is essentially the same as 2.0.1.7 but with the multithreading fixes from @FootStark enabled by default.

**Comment by jarmo** [comment]
Tried to run restore process with duplicati from `HEAD` like this:
```
restore "i:\backup.old\d" --restore-path="h:\restore\d" --dbpath="i:\backup.old\db\backup-db-d-old" --passphrase="xxx" --aes-set-threadlevel=4 --no-local-blocks="true" --debug-output="true" --verbose="true"
```

I'm getting an Exception:
```
{"Failed to execute SQL: \r\nDROP INDEX \"BlocksetEntryIds_Forward\"\nError: SQL logic error or missing database\r\nno such index: BlocksetEntryIds_Forward\nDatabase is NOT upgraded."}
```

Is the old database not compatible with new code or what's up with that? Also, should I set `--aes-set-threadlevel` to 8 if my CPU has 4 cores, but 8 threads?

**Comment by kenkendk** [comment]
The exception has something to do with switching database versions. I am not sure why it is attempting to update your database, but that depends on what version you ran previously.

The `--aes-set-threadlevel` cannot be more than 4. It will use at most 2 threads for encryption and at most 4 for decryption.

**Comment by FootStark** [comment]
@jarmo, it seems your db is broken. The new version does an update, that seemed to have failed. Did you kill ducplicati when the upgrade was running? There should be an automatic backup in your db-path, try to restore it and run again.

**Comment by jarmo** [comment]
I'm not sure why my db was broken, but I restored it from a backup and started the restore process. Now it seems to be working.

Here we go:
```
Searching backup 0 (13.10.2015 1:49.14) ...
Needs to restore 50641 files (766,85 GB)
```

Will look at the performance later.

**Comment by JF28** [comment]
Okay guys, in my case, Duplicati finally gave up this morning with this error message : 

```
Recreate completed, verifying the database consistency

System.IO.InvalidDataException: Found inconsistency in the following files while validating database:
xxx, actual size 345597906, dbsize 17917906, blocksetid: 2212
xxx, actual size 13121458002, dbsize 14258002, blocksetid: 3028
xxx, actual size 14765295058, dbsize 19695058, blocksetid: 3034
xxx, actual size 2643713106, dbsize22273106, blocksetid: 3058
xxx, actual size 346144338, dbsize 18464338, blocksetid: 64038
... and 24 more. Run repair to fix it.
   at Duplicati.Library.Main.Database.LocalDatabase.VerifyConsistency(IDbTransaction transaction, Int64 blocksize, Int64 hashsize)
   at Duplicati.Library.Main.Operation.RecreateDatabaseHandler.DoRun(LocalDatabase dbparent, Boolean updating, IFilter filter, NumberedFilterFilelistDelegate filelistfilter, BlockVolumePostProcessor blockprocessor)
   at Duplicati.Library.Main.Operation.RestoreHandler.Run(String[] paths, IFilter filter)
   at Duplicati.Library.Main.Controller.<Restore>c__AnonStorey1.<>m__0(RestoreResults result)
   at Duplicati.Library.Main.Controller.RunAction[T](T result, String[]& paths,IFilter& filter, Action`1 method)
   at Duplicati.Library.Main.Controller.Restore(String[] paths, IFilter filter)
   at Duplicati.CommandLine.Commands.Restore(List`1 args, Dictionary`2 options,IFilter filter)
   at Duplicati.CommandLine.Program.RealMain(String[] args)
```

Kenneth : there is something that does not add up here. I can assure you that **no** files have been modified or deleted manually on my backend. No dlist, dblock or dindex is missing, I am sure of it. I can not understand why Duplicati suddenly felt that something was corrupt on the backend... And ultimately failed while trying to repair the mess. Furthermore, all incremental backups have been successful so far (last incremental backup dates back to Friday... It is not so old !). So, on one side Duplicati tells me that the backups are running perfectly smoothly and everything is OK. On the other side, Duplicati tells me that nothing can be restored, something is broken somewhere and can not be repaired. And so, what ? All data on the backend should be disposed, and I have to restart from everything scratch ?! IMHO, this is a very disturbing behavior :(

I'll pick up my backup database and retry a restore with this it. I'll keep you informed.

**Comment by jarmo** [comment]
@JF28 maybe it's same to https://github.com/duplicati/duplicati/issues/1143 which I also stumbled on when trying to restore without a database. As soon as I gave database as a parameter everything went fine.

**Comment by FootStark** [comment]
@JF28: Yes, that looks like the issue happening when files with 320MB-chunks of unchanged data (full blocklist) are in the backup. I guess all 29 files mentioned are > 320MB? Kenneth was about to fix that.

**Comment by JF28** [comment]
jarno : indeed, I started a new full restore with the database, and everything seems to be going well so far. Yet, this failure is a very serious issue in my view. Duplicati seems so fragile without its beloved database.

FootStark : absolutely, all files listed are very large (> 10GB). FYI, restore with the database is ongoing... And it's as if Duplicati took steroids xD Restore speed is quite impressive. More information to come.

**Comment by jarmo** [comment]
It seems to me that my 766,85 GB have been restored hour or two ago, and currently it's integrity verification phase (with log statements like `Testing restored file integrity...`. So I can say that the restore took about ~5-6h - that's a huge improvement over ~14 days it took originally. Thank you very much @FootStark and @kenkendk for all the great work done by you so far!

I agree with @JF28 that it is a real problem when Duplicati is not able to restore anything when database is missing. After the verification step completes, I will try to restore everything again, but this time with `--no-local-db` option.

**Comment by JF28** [comment]
Here are my results (have to be compared with those of [February 18](https://github.com/duplicati/duplicati/issues/1391#issuecomment-185879872), it's nearly the same restore scenario) :

```
2016-03-21 17:37:25Z - Information: The operation Restore has started
2016-03-21 17:40:34Z - Information: Searching backup 0 (18-03-16 19:00:15) ...
2016-03-21 17:40:36Z - Information: Needs to restore 33679 files (175,51 GB)
2016-03-21 17:42:23Z - Profiling: Starting - ScanForexistingTargetBlocks
2016-03-21 17:42:43Z - Profiling: ScanForexistingTargetBlocks took 00:00:20.187
2016-03-21 17:42:43Z - Profiling: Starting - ScanForExistingSourceBlocksFast
2016-03-21 17:44:50Z - Profiling: ScanForExistingSourceBlocksFast took 00:02:06.877
2016-03-21 17:44:50Z - Profiling: Starting - GetMissingVolumes
2016-03-21 17:45:36Z - Profiling: GetMissingVolumes took 00:00:46.187
2016-03-21 17:45:36Z - Information: 2815 remote files are required to restore
2016-03-21 19:21:24Z - Profiling: Starting - RestoreVerification
2016-03-21 20:44:41Z - Profiling: RestoreVerification took 01:23:17.508
2016-03-21 20:47:40Z - Profiling: Starting - RemoteOperationTerminate
2016-03-21 20:47:40Z - Profiling: RemoteOperationTerminate took 00:00:00.000
```

In short :

- database rebuild... failed completely :(
- files restoration took ~96 minutes (**5 times faster** than previously, a stunning performance
 ^___^)
- restore verification took ~83 minutes (still room for improvement here ?)
- total restoration time : ~3 hours

Amazing... Duplicati really starts to rock :) Nice job FootStark & Kenneth !

**Comment by FootStark** [comment]
Glad to hear the performance of core restoration is now acceptable and stable for large backups, too. Other minor improvements may come, but I guess the db rebuild should get priority now.

**Comment by jarmo** [comment]
Unfortunately I cannot tell how much time went to the integrity verification, but it was completed when I woke up this morning. So, still a huge improvement compared to 14 days :)

Started restoration process with `--no-local-db` now and it has been doing it for last 4 hours without a single restored file yet. In logs I see `Downloading file...` lines.

**Comment by jarmo** [comment]
Ok, it has been working about 24h now and I can see that there is still not a single file restored. Not even empty directory tree has been created and in log I can see `Downloading file...` lines. When I pause this process with a debugger, I can see that it is in `FindMissingBlocklistHashes`. I'm going to give it additional 24h and then I'm going to kill it if there's no progress.

Still looks really sad situation to be in if there's no local database available.

**Comment by JF28** [comment]
It seems that your Duplicati, like mine, want to download all the dblock files just to rebuild the database (which turned out to be a complete waste of time in my case)

Given the last mishap, I changed my backup script, so that the database is automatically uploaded to the backend (in a 7z archive) after Duplicati has completed its backup job. Would be nice if Duplicati had an option to do it itself automatically.

**Comment by jarmo** [comment]
It is not downloading per-se, because everything is on a local physical disks. Uploading that database to a 3rd party destination might cause privacy concerns, because that database has information about your backup (file names, etc.). If that is not a problem, then go ahead, but I doubt that this should be a default way of Duplicati. As @FootStark has mentioned already then correct solution would be to fix/optimize database building task.

**Comment by JF28** [comment]
I also work with local disks for restore. "Download" is just a generic term for "file copy" in my mind, it doesn't necessarily imply a network traffic of any kind.

The 7z archive that contains my database is encrypted and protected by a password. Duplicati could do the same easily, without any privacy concerns.

**Comment by FootStark** [comment]
@jarmo As you're in the IDE, can you provide a callstack when pausing the application.
Furthermore, can you confirm that your backup also consists of files > 320MB and that you have multiple versions of them over incremental backups?
Maybe just stop the recreate for now then. From above (your takes-4-days test) I assume that database recreation was ok then (exactly same backup), is this correct? Maybe you can revert to 2.0.1.5 canary and check if the db restore was ok then.

**Comment by FootStark** [comment]
@JF28 Until the recreation Feature is fixed, you can also use the `SharpAESCrypt.exe` tool to encrypt your db when uploading it to the remote storage. It's actually the library duplicati uses too (for aes), and that would keep your file encryption consistent.

**Comment by jarmo** [comment]
Another problem with manually encrypting database and sending to some remote backup destination would be that you'd need to implement backend integration separately from duplicati. That's not a problem when local disk is used as a backup destination, but would be a bigger problem if for example Amazon would be used as a destination.

@JF28 hopefully answered all your questions to me below.

1) Here's that callstack:
```
 	[External Code]	
>	Duplicati.Library.Main.dll!Duplicati.Library.Main.Database.ExtensionMethods.ExecuteNonQuery(System.Data.IDbCommand self, string cmd, object[] values) Line 49	C#
 	Duplicati.Library.Main.dll!Duplicati.Library.Main.Database.LocalRecreateDatabase.FindMissingBlocklistHashes(long hashsize, long blocksize, System.Data.IDbTransaction transaction) Line 271	C#
 	Duplicati.Library.Main.dll!Duplicati.Library.Main.Operation.RecreateDatabaseHandler.DoRun(Duplicati.Library.Main.Database.LocalDatabase dbparent, bool updating, Duplicati.Library.Utility.IFilter filter, Duplicati.Library.Main.Operation.RecreateDatabaseHandler.NumberedFilterFilelistDelegate filelistfilter, Duplicati.Library.Main.Operation.RecreateDatabaseHandler.BlockVolumePostProcessor blockprocessor) Line 410	C#
 	Duplicati.Library.Main.dll!Duplicati.Library.Main.Operation.RestoreHandler.Run(string[] paths, Duplicati.Library.Utility.IFilter filter) Line 174	C#
 	Duplicati.Library.Main.dll!Duplicati.Library.Main.Controller.Restore.AnonymousMethod__0(Duplicati.Library.Main.RestoreResults result) Line 191	C#
 	Duplicati.Library.Main.dll!Duplicati.Library.Main.Controller.RunAction<Duplicati.Library.Main.RestoreResults>(Duplicati.Library.Main.RestoreResults result, ref string[] paths, ref Duplicati.Library.Utility.IFilter filter, System.Action<Duplicati.Library.Main.RestoreResults> method) Line 357	C#
 	Duplicati.Library.Main.dll!Duplicati.Library.Main.Controller.Restore(string[] paths, Duplicati.Library.Utility.IFilter filter) Line 190	C#
 	Duplicati.CommandLine.exe!Duplicati.CommandLine.Commands.Restore(System.Collections.Generic.List<string> args, System.Collections.Generic.Dictionary<string, string> options, Duplicati.Library.Utility.IFilter filter) Line 458	C#
 	Duplicati.CommandLine.exe!Duplicati.CommandLine.Program.RealMain(string[] args) Line 182	C#
 	[External Code]	
 	Duplicati.Library.AutoUpdater.exe!Duplicati.Library.AutoUpdater.UpdaterManager.RunMethod(System.Reflection.MethodInfo method, string[] args) Line 879	C#
 	Duplicati.Library.AutoUpdater.exe!Duplicati.Library.AutoUpdater.UpdaterManager.RunFromMostRecent.AnonymousMethod__0() Line 926	C#
 	Duplicati.Library.AutoUpdater.exe!Duplicati.Library.AutoUpdater.UpdaterManager.WrapWithUpdater(Duplicati.Library.AutoUpdater.AutoUpdateStrategy defaultstrategy, System.Action wrappedFunction) Line 843	C#
 	Duplicati.Library.AutoUpdater.exe!Duplicati.Library.AutoUpdater.UpdaterManager.RunFromMostRecent(System.Reflection.MethodInfo method, string[] cmdargs, Duplicati.Library.AutoUpdater.AutoUpdateStrategy defaultstrategy) Line 925	C#
 	Duplicati.CommandLine.exe!Duplicati.CommandLine.Program.Main(string[] args) Line 36	C#
 	[External Code]	
 	Duplicati.Library.AutoUpdater.exe!Duplicati.Library.AutoUpdater.UpdaterManager.RunFromMostRecent(System.Reflection.MethodInfo method, string[] cmdargs, Duplicati.Library.AutoUpdater.AutoUpdateStrategy defaultstrategy) Line 989	C#
 	Duplicati.CommandLine.exe!Duplicati.CommandLine.Program.Main(string[] args) Line 36	C#
 	[External Code]	
```

2) Yes, there probably are files larger than 320MB and it's highly likely that they have incremental backups.

3) Database is exactly the same, but recreation failed in the past - only my 50Gb backup succeeded to recreate it's database and luckily that backup included database for that 700Gb backup as well which eventually helped me out.

4) Can you give me git commit hash for reverting. It would be easier to do everything through IDE and get better information when needed (like that callstack above, for example).

**Comment by FootStark** [comment]
Thanks for the info.
Commit would be: 9e94e05b06f8365f127d4e40ea9ca4016ebc3501
But if it didn't work before, you can spare the effort.

**Comment by jarmo** [comment]
Okay, about 48h situation is the same (even stacktrace seems to be the same) and not a single file is restored. I killed it and not try again until further optimizations.

Thanks for everything so far!

**Comment by jorisvd** [comment]
Thanks for all your contributions! I've just read through the complete thread here with great interest. Could someone tell me whether the initial slow repair problem was also affecting Duplicati v1? Does that also use a SQLite DB and where would it typically be stored (I have been looking in the AppData folders but cannot find it...)

**Comment by renestach** [comment]
This is only about Duplicati 2.0.

**Comment by jarmo** [comment]
Even worse situation happened in #2140 where restoration fails completely without a local database.

**Comment by kenkendk** [comment]
I am collecting performance related issues under #2302
